### PR TITLE
preliminary world chunking and example

### PR DESCRIPTION
**Pull request by JtotheThree**
Loads World files with my rs-tiled fork. Before we open the PR there, I'd like to test fully loading the tmx files from rs-tiled based on the world file and see how that performs.

TiledWorldSettings can be set to chunk based on dimensions around the center of the camera 800x800 for example. A system is added to handle the chunking, if a maps Rect falls inside the chunking rect then it's loaded. Once it falls out it's unloaded.

I have added example/world.rs demonstrating this.

I'm opening this PR as preliminary if there's any notes on directions (resource vs component for the settings for instance) before proceeding futher.

### CODE REVIEW

File: src/world/mod.rs [hunk 1]
```diff
@@ -1,16 +1,99 @@
...
+
+pub(crate) fn world_chunking(
+    camera: Query<&Transform, (With<Camera>, Changed<Transform>)>,    
+    worlds: Res<Assets<TiledWorld>>,
+    mut commands: Commands,
+    asset_server: Res<AssetServer>,
+    mut world_query: Query<(Entity, &TiledWorldHandle, &TiledWorldSettings)>,
+    mut chunked_maps: ResMut<WorldChunkedMaps>,
+) {
+    let camera_transform = match camera.iter().next() {
```
Comment (adrien-bon) [inline src/world/mod.rs:25]: What if we have several cameras ?
I think we should iterate over all of them, so we can construct a list of "maps to remove" and a list of "maps to add" then do the actually spawn / despawn.

**Review by adrien-bon — COMMENTED** [review]
Thanks for your contribution! :)

Please see my comments for details.
To sum up most important ones:
- I'd prefer to store maps entities directly in a world component instead of a resource
- Preload maps asset during world asset loading. Should be OK performance-wise since it's actually only parsing text files
- Handle several cameras for the chunking logic

I only reviewed your change and did not have the chance to actually test them so I may have missed some things.

Anyway, keep up the good work ! :)

File: src/world/mod.rs [hunk 2]
```diff
@@ -1,16 +1,106 @@
...
+        let offset = match tiled_settings.layer_positioning {
+            LayerPositioning::Centered => Vec3::new(
+                -tiled_world.world_rect.max.x as f32 / 2.0,
+                -tiled_world.world_rect.max.y as f32 / 2.0,
+                0.0,
+            ),
+            LayerPositioning::TiledOffset => Vec3::ZERO,
+        };
+        
+        let tiled_world = worlds.get(&world_handle.0).unwrap();
```
Comment (adrien-bon) [inline src/world/mod.rs:47]: Encountered a panic on this line: we could reach this piece of code while the world is not yet fukly loaded.
Please change this to `let Some(tiled_world) = ... else { return; };`

File: Cargo.toml [hunk 3]
```diff
@@ -48,14 +48,13 @@ user_properties = []
...
-tiled = "0.13"
+# tiled = "0.13"
+tiled = { git = "https://github.com/JtoTheThree/rs-tiled.git", branch = "worldfiles" }
 
 # Utilities
 futures-lite = "2.5"
 log = "0.4"
 thiserror = "1.0"
-serde = { version = "1.0", features = ["derive"] }
-serde_json = { version = "1.0" }
```
[the remark this answers is not in the document]
Comment (JtotheThree) [reply Cargo.toml:58]: I must have accidentally changed cargo.toml here when I was working on the rs-tiled PR. Resolved.

File: examples/world.rs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+            TextFont {
+                font_size: 24.0,
+                ..default()
+            },
+            HelperText,
+        ));
+
+}
+
+fn input(
```
[the remark this answers is not in the document]
Comment (JtotheThree) [reply examples/world.rs]: Updated to use the HelperPlugin.

File: src/world/asset.rs [hunk 5]
```diff
@@ -6,7 +7,7 @@ use bevy::{asset::{io::Reader, AssetLoader, LoadContext}, prelude::*};
 /// `Asset` holding Tiled world informations.
 #[derive(TypePath, Asset)]
 pub struct TiledWorld {
```
[the remark this answers is not in the document]
Comment (JtotheThree) [reply src/world/asset.rs:13]: Handles are stored now. I've tested with my other crate and it's needed anyway if the user wants to pull any information from the maps if they're not chunked in.

File: src/world/components.rs [hunk 6]
```diff
@@ -4,7 +4,9 @@ use bevy::prelude::*;
 
 #[derive(Component, Default)]
 pub struct TiledWorldSettings {
-    // XXX: here we could have world related settings, such as culling
+    pub chunking: bool,
+    pub chunking_width: u32,
+    pub chunking_height: u32,
```
[the remark this answers is not in the document]
Comment (JtotheThree) [reply src/world/components.rs]: Switched to Option<(u32, u32)>

File: src/world/mod.rs [hunk 7]
```diff
@@ -1,16 +1,99 @@
...
+            {
+                continue;
+            }
+
+            let map_path = world_path.path().parent().expect("Parent path not found").join(map_path.clone());
+
+            let map_entity = commands
+                .spawn_empty()
+                .insert(TiledMapHandle(asset_server.load(map_path.clone())))
+                .insert(RespawnTiledMap)
```
[the remark this answers is not in the document]
Comment (JtotheThree) [reply src/world/mod.rs]: Removed and tested.

File: examples/world.rs [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+        TiledMapSettings {
+            layer_positioning: LayerPositioning::Centered,
+            ..default()
+        },
+        TiledWorldSettings {
+            chunking: true,
+            chunking_width: 200,
+            chunking_height: 200,
+        },
+        TilemapRenderSettings::default(),
```
[the remark this answers is not in the document]
Comment (JtotheThree) [reply examples/world.rs]: Removed.

File: src/world/mod.rs (hunk 1 above)
Comment (JtotheThree) [reply src/world/mod.rs]: Added handling for multiple cameras.

File: src/world/mod.rs (hunk 2 above)
Comment (JtotheThree) [reply src/world/mod.rs]: Added the option test.

File: src/world/mod.rs [hunk 9]
```diff
@@ -57,7 +134,25 @@ pub(crate) fn process_loaded_worlds(
             // Clean world
             remove_maps(&mut commands, &mut world_storage);
 
-            // XXX: load the world, ie. update TiledWorldStorage component based upon informations from TiledWorldHandle / TiledWorld
+            if world_settings.chunking {
+                return;                
+            }
+
+            // Get the relative path of the world asset 
+            let world_parent_path = world_handle.0.path().unwrap().parent().expect("World parent path not found");
```
[the remark this answers is not in the document]
Comment (JtotheThree) [reply src/world/mod.rs]: Map asset loading is now in assets.rs and stores the handles for later use.

**Comment by adrien-bon** [comment]
Merging this PR as is, lets continue the work over at #63

**Pull request merged**